### PR TITLE
Makefile: move SSL options into a block and refine rules

### PR DESCRIPTION
**Pull request by pizhenwei**
Makefile: move SSL options into a block and refine rules
    
    Move SSL options into a block to make it easy to read, and change
    global rules to make it easy to maintain. For the further step, it
    gets extensible to add another type.
    
    New version of the library building rule:
            static: $(STLIBNAME) $(SSL_STLIB)
            dynamic: $(DYLIBNAME) $(SSL_DYLIB)
    
    Compare with the orignal version:
            dynamic: $(DYLIBNAME)
            static: $(STLIBNAME)
            ifeq ($(USE_SSL),1)
            dynamic: $(SSL_DYLIBNAME)
            static: $(SSL_STLIBNAME)
            endif
    
    If we want to add a new type(Ex, RDMA), for the new version, we can
    do like this:
            ########### RDMA options start ###############
            .....
            ########### RDMA options end   ###############
    
            static: $(STLIBNAME) $(SSL_STLIB) $(RDMA_STLIB)
            dynamic: $(DYLIBNAME) $(SSL_DYLIB) $(RDMA_DYLIB)
    
    Signed-off-by: zhenwei pi <pizhenwei@bytedance.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -60,22 +55,38 @@ DYLIB_MAKE_CMD=$(CC) -shared -Wl,-soname,$(DYLIB_MINOR_NAME)
 STLIBNAME=$(LIBNAME).$(STLIBSUFFIX)
 STLIB_MAKE_CMD=$(AR) rcs
 
+uname_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
+
+# Deps (use make dep to generate this)
+alloc.o: alloc.c fmacros.h alloc.h
```
Comment (michael-grunder) [inline Makefile:61]: I think this will need to be moved below the `all:` target since right now running make without args only builds `alloc.o`

**Comment by michael-grunder** [comment]
Hi, thanks for the PR.

If you rebase your branch against master (and move the deps targets) I will test and get this merged.

**Comment by pizhenwei** [comment]
> Hi, thanks for the PR.
> 
> If you rebase your branch against master (and move the deps targets) I will test and get this merged.

Force updated, thanks for your suggestion.

**Comment by michael-grunder** [comment]
Looks like this breaks the tests in Windows, because it's unconditionally adding `-lssl -lcrypto -lpthread` regardless of the `USE_SSL` value.

This [small modification on my fork](https://github.com/michael-grunder/hiredis/commit/89f5c454ac5748319118ac4243ea7766800fe649) seems to solve the problem.

If that looks OK to you I can include the patch when I merge.

**Comment by pizhenwei** [comment]
> Looks like this breaks the tests in Windows, because it's unconditionally adding `-lssl -lcrypto -lpthread` regardless of the `USE_SSL` value.
> 
> This [small modification on my fork](https://github.com/michael-grunder/hiredis/commit/89f5c454ac5748319118ac4243ea7766800fe649) seems to solve the problem.
> 
> If that looks OK to you I can include the patch when I merge.

Thanks!

**Comment by michael-grunder** [comment]
Merged, thanks!

**Pull request merged**